tools/tungsten/core/src/com/google/android/filament/tungsten/texture/TextureUtils.kt

```diff
@@ -0,0 +1,137 @@
+ * limitations under the License.
+ */
+
+package com.google.android.filament.tungsten.texture
+
+import com.google.android.filament.Engine
+import com.google.android.filament.Texture
+
+import javax.imageio.ImageIO
+import java.awt.image.BufferedImage
+import java.awt.image.DataBufferByte
+import java.io.File
+import java.io.IOException
+import java.nio.ByteBuffer
+import java.nio.ByteOrder
+
+private fun Texture.InternalFormat.toSrgb(): Texture.InternalFormat {
+    return when (this) {
+        Texture.InternalFormat.RGB8 -> Texture.InternalFormat.SRGB8
+        Texture.InternalFormat.RGBA8 -> Texture.InternalFormat.SRGB8_A8
+        else -> this
+    }
+}
+
+object TextureUtils {
+
+    fun createDefaultTexture(engine: Engine): Texture {
+        val stream = javaClass.classLoader.getResourceAsStream("checkerboard.png")
+        val texture = stream?.let { s ->
+            ImageIO.read(s)?.let {
+                img -> loadTextureFromImage(engine, img)
+            }
+        }
+        return texture ?: throw RuntimeException("Could not load default texture")
+    }
+
+    fun loadTexture(engine: Engine, file: File): Texture? {
+        val img = try {
+            ImageIO.read(file) ?: return null
+        } catch (e: IOException) {
+            System.err.println("Could not read image ${file.canonicalPath}.")
+            e.printStackTrace()
+            return null
+        }
+
+        return loadTextureFromImage(engine, img)
+    }
+
+    /**
+     * Based on the number of components the image has and its color space, decide which
+     * texture formats to use.
+     */
+    private fun decideTextureFormat(
+        img: BufferedImage
+    ): Pair<Texture.InternalFormat, Texture.Format> {
+        val isSrgb = img.colorModel.colorSpace.isCS_sRGB
+        val (internalFormat, textureFormat) = when (img.colorModel.numComponents) {
+            1 -> Texture.InternalFormat.R8 to Texture.Format.R
+            2 -> Texture.InternalFormat.RG8 to Texture.Format.RG
+            3 -> Texture.InternalFormat.RGB8 to Texture.Format.RGB
+            4 -> Texture.InternalFormat.RGBA8 to Texture.Format.RGBA
+            else -> Texture.InternalFormat.RGBA8 to Texture.Format.RGBA
+        }
+        return (if (isSrgb) internalFormat.toSrgb() else internalFormat) to textureFormat
+    }
+
+    private fun loadTextureFromImage(engine: Engine, img: BufferedImage): Texture? {
+        val data = img.data.dataBuffer as DataBufferByte
+        val pixels = data.data
+
+        val (internalFormat, textureFormat) = decideTextureFormat(img)
+
+        flipComponents(img.colorModel.numComponents, pixels)
+
+        val texture = Texture.Builder()
+                .width(img.width)
+                .height(img.height)
+                .format(internalFormat)
+                .sampler(Texture.Sampler.SAMPLER_2D)
+                .build(engine)
+
+        val buf = ByteBuffer.wrap(pixels)
+        buf.order(ByteOrder.BIG_ENDIAN)
```

Comment:
You could probably play with the byte order here to avoid flipComponents or at least make it faster

Reply:
Calling `.order` here is actually useless- it only affects the interpretation of the buffer when doing calls like `.getInt`. I wasn't able to find a better way to swap the components, but I'll keep looking.

Reply:
Yeah I guess the problem is that our JNI doesn't look at the order and expects native order (little endian). We could make the JNI do this work, but this is fine for now.